library/core/src/array/mod.rs

```diff
@@ -606,8 +608,20 @@ impl<T, const N: usize> [T; N] {
     /// assert_eq!(strings.len(), 3);
     /// ```
     #[stable(feature = "array_methods", since = "1.77.0")]
-    pub fn each_ref(&self) -> [&T; N] {
-        from_trusted_iterator(self.iter())
+    #[rustc_const_unstable(feature = "const_array_each_ref", issue = "133289")]
+    pub const fn each_ref(&self) -> [&T; N] {
+        let mut buf = [null::<T>(); N];
+
+        // FIXME(const-hack): We would like to simply use iterators for this (as in the original implementation), but this is not allowed in constant expressions.
+        let mut i = 0;
+        while i < N {
+            buf[i] = &raw const self[i];
+
+            i += 1;
+        }
+
+        // SAFETY: `*const T` has the same layout as `&T`, and we've also initialised each pointer as a valid reference.
+        unsafe { transmute_unchecked(buf) }
```

Comment:
What is the need for using the intrinsic over using `core::mem::transmute` here? The sizes should be the same, so bypassing the check shouldn't be necessary as far as I can tell.

Reply:
`transmute` does not allow arrays with generic bounds (or dependently-sized types, as the compiler calls them).

Example: <https://godbolt.org/z/86Gfnrd66>

Reply:
Gotcha. Hopefully that can be lifted in the future!